### ISSUE TITLE
How are the shape functions built?

### ISSUE THREAD

**Issue by kensen10** [interpretml/interpret#114]
I want to have a clear understanding of how the scores from each feature is derived. I assume it's derived from a tree-ensembled model.

If that's the case, are the shape functions model (tree-ensembled models) built sequentially or at the same time? Let's say we have 20 features, each feature has a tree ensembled model itself, but I couldn't find detail about whether the individual model is built sequentially or independently or at the same time.

Does each shape function model share the same target variable?

Thanks!

**Comment by interpret-ml** [comment]
Hi @kensen10, 

You're correct, the shape functions are derived from an ensemble of trees, each of which is restricted to a single feature. Here's a high level description of the algorithm:

These shape functions learned _cyclically_. For a model with 20 features, we will first learn a shallow tree on the first feature f<sub>0</sub>. This shallow tree is penalized by a learning rate (typically 0.01) to ensure that the ordering of features doesn't play a significant role. We then update the residuals, and move onto the next feature f<sub>1</sub>, which learns its own shallow tree on the residuals (not the original targets). This repeats until we finally learn a tree for f<sub>20</sub>, at which point we cycle back to f<sub>0</sub> and repeat the process. Finally, after a fixed number of epochs or if an early stopping threshold is met, a model is produced. 

Ultimately, we train many of these models on different samples of the data (defaulted to 16 with the n_estimators parameter), and average these models together to produce the final shape functions. This averaging of bagged models produces smoother, more intelligible graphs and is what we use to determine uncertainty intervals.

We're happy to clarify more if you have further questions! The papers in our citation section on the readme also go into much more detail. In particular, we'd recommend reading the original papers describing the algorithm (https://www.cs.cornell.edu/~yinlou/papers/lou-kdd12.pdf ) and (http://www.cs.cornell.edu/~yinlou/papers/lou-kdd13.pdf ). A light description and applied use case can be found here (https://www.microsoft.com/en-us/research/wp-content/uploads/2017/06/KDD2015FinalDraftIntelligibleModels4HealthCare_igt143e-caruanaA.pdf ).

-InterpretML Team

**Comment by kensen10** [comment]
Thank you! That answers my question :)
